Src/HALALMock/Services/PWM/PWM/PWM.cpp

```diff
@@ -1,110 +1,63 @@
-
+#include "HALALMock/Services/PWM/PWM/PWM.hpp"
 PWM::PWM(Pin& pin) {
-	if (not TimerPeripheral::available_pwm.contains(pin)) {
-
-		ErrorHandler("Pin %s is not registered as an available PWM", pin.to_string());
-		return;
-	}
-
-	TimerPeripheral& timer = TimerPeripheral::available_pwm.at(pin).first;
-	TimerPeripheral::PWMData& pwm_data = TimerPeripheral::available_pwm.at(pin).second;
-
-	if (pwm_data.mode != TimerPeripheral::PWM_MODE::NORMAL) {
-		ErrorHandler("Pin %s is not registered as a NORMAL PWM", pin.to_string());
+	/*By HalalMock we only will test the logical interface of PWM
+	if the PWM works, however when you try in your micro doesn't work,
+	you should check the timers and channels, that could be the problem 
+	*/
+	EmulatedPin &pin_memory=SharedMemory::get_pin(pin);
+	if(pin_memory.type==PinType::NOT_USED){
+		pin_memory.type=PinType::PWM;
+		pin_memory.PinData.PWM.duty_cycle=0.0f;
+		pin_memory.PinData.PWM.is_on=false;
+		//let's point our class variables to the variables from PinModel
+		this->duty_cycle=&(pin_memory.PinData.PWM.duty_cycle);
+		this->frequency=&(pin_memory.PinData.PWM.frequency);
+		this->is_on=&(pin_memory.PinData.PWM.is_on);
+		//default values
+		*(this->duty_cycle)=0.0f;
+		*(this->is_on)=false;
+		*(this->frequency)=0;
+	}else{
+		ErrorHandler("Pin %s is being used already",pin.to_string());
 	}
-
-	peripheral = &timer;
-	channel = pwm_data.channel;
-
-	Pin::inscribe(pin, TIMER_ALTERNATE_FUNCTION);
-	timer.init_data.pwm_channels.push_back(pwm_data);
-
-	duty_cycle = 0;
-	is_initialized = true;
 }
 
 void PWM::turn_on() {
-  if(not is_initialized){
-	ErrorHandler("PWM was not initialized");
-  }
-  if(is_on) return;
-  if(HAL_TIM_PWM_Start(peripheral->handle, channel) != HAL_OK) {
-    ErrorHandler("PWM did not turn on correctly", 0);
-  }
-  is_on = true;
+	if(*(this->frequency)==0){
+		ErrorHandler("You need to set a frequency before turn_on the PWM :%d",*(this->frequency));
+	}else{
+		*(this->is_on) = true;
+	}
+  
 }
 
 void PWM::turn_off() {
-  if(not is_on) return;
-  if(HAL_TIM_PWM_Stop(peripheral->handle, channel) != HAL_OK) {
-    ErrorHandler("PWM did not stop correctly", 0);
-  }
-  is_on = false;
+  *(this->is_on) = false;
 }
 
 void PWM::set_duty_cycle(float duty_cycle) {
-	uint16_t raw_duty = __HAL_TIM_GET_AUTORELOAD(peripheral->handle) / 100.0 * duty_cycle;
-	__HAL_TIM_SET_COMPARE(peripheral->handle, channel, raw_duty);
-	this->duty_cycle = duty_cycle;
+	*(this->duty_cycle) = duty_cycle;
 }
 
 void PWM::set_frequency(uint32_t frequency) {
-	this->frequency = frequency;
-	TIM_TypeDef& timer = *peripheral->handle->Instance;
-	timer.ARR = (HAL_RCC_GetPCLK1Freq()*2 / (timer.PSC+1)) / frequency;
-	set_duty_cycle(duty_cycle);
+	*(this->frequency) = frequency;
 }
 
 uint32_t PWM::get_frequency() {
-	return frequency;
+	return *(this->frequency);
 }
 
 float PWM::get_duty_cycle(){
-	return duty_cycle;
+	return *(this->duty_cycle);
 }
 void PWM::set_dead_time(std::chrono::nanoseconds dead_time_ns)
```

Comment:
I would add a new variable to the union that represents the dead time in nanoseconds, it is easy and also important to see if the deadtime is being configured